### PR TITLE
Initial commit of Multi Token Prediction

### PR DESCRIPTION
**Pull request by parambole**
# Description

This change introduces the scaffolding for Multi-Token Prediction into the Maxtext framework, based on the methodology described in the Deepseek V3 paper (https://arxiv.org/pdf/2412.19437).

Specifically, it adds the MultiTokenPredictionLayer, containing an initial implementation of the forward pass. This implementation covers equations (21), (22), and (23) detailed in the paper.

# Tests

A new unit test suite (`multi_token_predicton_test.py`) was added to verify the `MultiTokenPredictionLayer`. The tests ensure:

- Does the layer run without crashing when given inputs of the expected shape and type?
- Does the layer produce an output tensor with the correct dimensions, matching the expected hidden dimension?
- Does the output contain any NaN (Not a Number) or Inf (Infinity) values, which would indicate a problem?

The tests were executed successfully using the following command:

```
python -m pytest multi_token_predicton_test.py -s -v
```

output:
```
MultiTokenPredictionLayer test passed!
  Config Batch: 12, SeqLen: 2048, EmbedDim: 2048
  Input shape (hidden): (12, 2048, 2048)
  Input shape (target_emb): (12, 2048, 2048)
  Output shape: (12, 2048, 2048)
PASSED

========================================================================= 1 passed in 20.01s ==========================================================================
```

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

**Review by mailvijayasingh — APPROVED** [review]

File: MaxText/layers/multi_token_prediction.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+      2. Concatenation and Projection of normalized features.
+      3. Processing through a Transformer Decoder Layer.
+
+      Equation Representation (Conceptual):
+          norm_h = RMSNorm(h_prev)
+          norm_e = RMSNorm(e_target)
+          h_proj = W_p(concat(norm_h, norm_e))
+          h_next = InternalTransformerLayer(h_proj, pos_ids, segment_ids, ...)
+
+      It takes the previous state and target embedding as input and outputs the
```
Comment (gobbleturk) [inline MaxText/layers/multi_token_prediction.py:50]: previous state == previous hidden representation?

File: MaxText/layers/multi_token_prediction.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+      deterministic: bool,
+      model_mode: str = common_types.MODEL_MODE_TRAIN,
+  ) -> jnp.ndarray:
+    """
+    Applies the MTP combination, projection, and internal transformer processing.
+
+    Args:
+        prev_hidden_state: Hidden state from the previous step/layer.
+                           Shape: [batch, seq_len, hidden_size]
+        target_token_embedding: Embedding of the target token (e.g., token t+k).
```
Comment (gobbleturk) [inline MaxText/layers/multi_token_prediction.py:75]: nit: "token t+k" doesn't make sense to readers yet as k is not defined, I see below k is layer_number

File: MaxText/tests/multi_token_predicton_test.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    )
+
+    # Check dtype
+    self.assertEqual(
+        output_hidden_state.dtype,
+        self.cfg.dtype,
+        f"Expected output dtype {self.cfg.dtype}, but got {output_hidden_state.dtype}",
+    )
+
+    # Check for NaNs/Infs
```
Comment (gobbleturk) [inline MaxText/tests/multi_token_predicton_test.py:106]: I remember you asking why we don't check these in our other tests =)

File: MaxText/layers/multi_token_prediction.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+Config = common_types.Config
+DType = common_types.DType
+Mesh = common_types.Mesh
+RMSNorm = normalizations.RMSNorm
+DenseGeneral = linears.DenseGeneral
+TransformerLayerModuleType = models.DecoderLayer
+
+
+class MultiTokenPredictionLayer(nn.Module):
+  """
```
Comment (gobbleturk) [inline MaxText/layers/multi_token_prediction.py:38]: Awesome documentation!

File: MaxText/tests/multi_token_predicton_test.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    init_rngs = {"params": init_rng}
+    self.variables = self.mtp_layer.init(
+        init_rngs,
+        self.prev_hidden_state,
+        self.target_token_embedding,
+        self.position_ids,
+        self.decoder_segment_ids,
+        deterministic=True,
+    )
+    print("Setup complete.")
```
Comment (gobbleturk) [inline MaxText/tests/multi_token_predicton_test.py:76]: nit: remove or use maxtext logging, e.g.

`max_logging.log("Setup complete.")`

**Review by gobbleturk — APPROVED** [review]
Awesome documentation and test, thanks Param! I left some nits

File: MaxText/layers/multi_token_prediction.py (hunk 1 above)
Comment (parambole) [reply MaxText/layers/multi_token_prediction.py]: Yes - I have modified the comment.

File: MaxText/layers/multi_token_prediction.py (hunk 2 above)
Comment (parambole) [reply MaxText/layers/multi_token_prediction.py]: I have added more context thanks for pointing that out.

File: MaxText/layers/multi_token_prediction.py (hunk 4 above)
Comment (parambole) [reply MaxText/layers/multi_token_prediction.py:38]: Thank you :)

File: MaxText/tests/multi_token_predicton_test.py (hunk 5 above)
Comment (parambole) [reply MaxText/tests/multi_token_predicton_test.py]: Changed to `max_logging.log`

File: MaxText/layers/multi_token_prediction.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+    Returns:
+        next_hidden_state: The hidden state produced by this MTP step's internal transformer.
+                           Shape: [batch, seq_len, hidden_size]
+    """
+    cfg = self.config
+    mesh = self.mesh
+    k = self.layer_number
+
+    # --- 1. Normalize Hidden State and Embedding ---
+
```
Comment (gagika) [inline MaxText/layers/multi_token_prediction.py:96]: nit: Do you need the blank line?

File: MaxText/tests/multi_token_predicton_test.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+from MaxText import max_logging, pyconfig
+from MaxText import max_utils
+from MaxText.globals import PKG_DIR
+from MaxText.layers import models
+from MaxText.layers import multi_token_prediction  # The class under test
+
+TEST_LAYER_NUM = 1
+
+
+class MultiTokenPrediction(unittest.TestCase):
```
Comment (gagika) [inline MaxText/tests/multi_token_predicton_test.py:32]: MultiTokenPredictionLayerTest ?

File: MaxText/tests/multi_token_predicton_test.py
Comment (gagika) [inline MaxText/tests/multi_token_predicton_test.py:1]: typo in the filename: multi_token_predicton_test.py -> multi_token_prediction_test.py

**Review by gagika — APPROVED** [review]
Thanks Param!
I have only minor comments

File: MaxText/tests/multi_token_predicton_test.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+    # Prepare Dummy Input Data
+    prev_hidden_state_shape = (self.batch_size, self.seq_len, self.embed_dim)
+    target_embedding_shape = (self.batch_size, self.seq_len, self.embed_dim)
+    data_rng1, data_rng2, init_rng = jax.random.split(self.rng, 3)
+
+    self.prev_hidden_state = jax.random.normal(data_rng1, prev_hidden_state_shape, dtype=self.cfg.dtype)
+    self.target_token_embedding = jax.random.normal(data_rng2, target_embedding_shape, dtype=self.cfg.dtype)
+    self.position_ids = jnp.arange(self.seq_len, dtype=jnp.int32).reshape(1, -1).repeat(self.batch_size, axis=0)
+    self.decoder_segment_ids = jnp.ones((self.batch_size, self.seq_len), dtype=jnp.int32)
```
Comment (wang2yn84) [inline MaxText/tests/multi_token_predicton_test.py:63]: I think decoder_segment_id always starts from 0?

**Review by wang2yn84 — APPROVED** [review]
Overall looks good. Thank you!

File: MaxText/tests/multi_token_predicton_test.py
Comment (parambole) [reply MaxText/tests/multi_token_predicton_test.py:1]: Nice catch :) Took me some time to observe what was incorrect. Thanks. 

Corrected

File: MaxText/tests/multi_token_predicton_test.py (hunk 8 above)
Comment (parambole) [reply MaxText/tests/multi_token_predicton_test.py:63]: Hey @wang2yn84 - IIUC `decoder_segment_id` is used to handle padding i.e. for calculating the attention mask for sequences with unequal lengths. Over here I am trying to just simulate a very simple case with no padding. 

I have added a comment please let me know if I am missing something.

File: MaxText/layers/multi_token_prediction.py (hunk 6 above)
Comment (parambole) [reply MaxText/layers/multi_token_prediction.py]: Corrected.

File: MaxText/layers/multi_token_prediction.py [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+        dtype=cfg.dtype,
+        weight_dtype=cfg.weight_dtype,
+        kernel_axes=("concat_embed", "embed"),
+        name=f"mtp_{k}_projection",
+    )
+    # Shape: [B, S, H]
+    projected_features = projection_layer(concatenated_features)
+
+    # --- 4. Pass through MTP Transformer Block ---
+    next_hidden_state, _ = self.transformer_layer_module(config=cfg, mesh=mesh, name=f"mtp_{k}_transformer_layer")(
```
Comment (RissyRan) [inline MaxText/layers/multi_token_prediction.py:132]: I think this part is similar with [here](https://github.com/AI-Hypercomputer/maxtext/blob/da3acbd25eccce2f9e27e191bf53b6944ca7fe20/MaxText/layers/models.py#L456-L526 in models.py. I am not sure if we will meet some issues when integrate with each model, probably it's fine. The config usually include all layers.

File: MaxText/tests/multi_token_prediction_test.py [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+        self.target_token_embedding,
+        self.position_ids,
+        decoder_segment_ids=self.decoder_segment_ids,
+        deterministic=True,
+    )
+    # Assertions using unittest methods
+    expected_output_shape = (self.batch_size, self.seq_len, self.embed_dim)
+
+    # Check shape
+    self.assertEqual(
```
[the remark this answers is not in the document]
Comment (parambole) [reply MaxText/tests/multi_token_prediction_test.py:93]: Hey @RissyRan - Yes currently we are just checking the shapes as the forward pass is for scaffolding to verify the equations with respect to the shapes. The plan is to include more tests once we have the other components in.

File: MaxText/layers/multi_token_prediction.py (hunk 9 above)
Comment (parambole) [reply MaxText/layers/multi_token_prediction.py:132]: @RissyRan - Right now it's supposed to be layer agnostic and only depend on the previous hidden state and the current emb.
